README.md
# Leaflet Layer OverPass

This plugin is a fork of https://github.com/kartenkarsten/leaflet-layer-overpass

Here is a list of changes made after the fork:

* Replace `{{bbox}}` instead of `BBOX` in Overpass requests
* Code cleanup

 * JSHint warnings fixed
 * Tabs instead of spaces for indentation
 * Spaces between blocks

* Options are now in camelCase: `endpoint` is `endPoint` and so on
* Rename the `callback` option to `onSuccess`
* Add the `timeout` option
* Add the `onError` and `onTimeout` callbacks
* Add the ability to retry a request when it returned a timeout
* If beforeRequest return false, the request is not launched
* Add the noInitialRequest option to prevent the initial request to be launched if necessary
* Adds some significant refactor of code.
* Removes the requestPerTile mode.
* Extends the requests bounds.
* Adds a debug mode which shows the requests bounds.
* Invalidates the requests cache on error.

The first release of this fork is the version **1.1.0**.


## What is it?
A [Leaflet](http://leafletjs.com/) plugin to create a custom POI overlay - thanks to the [OSM](http://www.openstreetmap.org/) dataset and the [Overpass API](http://overpass-api.de/)

checkout the [demo](http://kartenkarsten.github.io/leaflet-layer-overpass/demo/)


## Installation
You can use bower to install leaflet-layer-overpass.

Simply run
```bash
$ bower install --save leaflet-layer-overpass
```
After that you can include and use the `OverpassLayer.css` and `OverpassLayer.js` files (or `OverPassLayer.min.js` if you want the minified version) from the `dist` folder in your html.

## How to use it?
```javascript
var attr_osm = 'Map data &copy; <a href="http://openstreetmap.org/">OpenStreetMap</a> contributors',

attr_overpass = 'POI via <a href="http://www.overpass-api.de/">Overpass API</a>',

osm = new L.TileLayer('http://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png', {

    'opacity': 0.7,
    'attribution': [attr_osm, attr_overpass].join(', ')
}),

map = new L.Map('map')
.addLayer(osm)
.setView(new L.LatLng(52.265, 10.524), 14),

opl = new L.OverPassLayer({

    query: 'node({{bbox}})['amenity'='post_box'];out;',
});

map.addLayer(opl);
```
In order to get a valid query the [Overpass-turbo IDE](http://overpass-turbo.eu/) might help.

## What are the options?
You can specify an options object as an argument of L.OverPassLayer.
```javascript
options: {

  'minZoom': 15,
  'requestPerTile': true,
  'endPoint': 'http://overpass.osm.rambler.ru/cgi/',
  'query': 'node({{bbox}})["amenity"="post_box"];out;',
  'timeout': 30 * 1000, // Milliseconds
  'minZoomIndicatorOptions': {

    'position': 'topright',
    'minZoomMessageNoLayer': 'no layer assigned',
    'minZoomMessage': 'current Zoom-Level: CURRENTZOOM all data at Level: MINZOOMLEVEL'
},
  'beforeRequest': function() {},
  'afterRequest': function() {},
  'onSuccess': function(data) {},
  'onError': function(xhr) {},
  'onTimeout': function(xhr) {},
};
```

## Used by
- [Gdzie](http://gdzie.bl.ee/#!7/51.495/20.995/)
- [pois.elblogdehumitos.com.ar](http://pois.elblogdehumitos.com.ar/)
- [briefkastenkarte.de](http://briefkastenkarte.de/)

## Dependencies
- Leaflet (tried with version 0.6.2, 0.7.3)

## Development

*Warning: This fork use Git Flow to manage branches.*

In order to contribute to the project you should first clone the repository. The javascript source files
reside in the `src` folder and are concatenated and minified by gulp. If you want to make changes
make them in the `src` folder and then build the `dist` file with gulp.
For that you first need to install gulp if you do not have installed it yet
```
$ npm install --global gulp
```
Then install all the needed packages for this project:
```
$ npm install
```
And then just run
```
gulp
```
after you made your changes. This will combine (and minify) the files and put them into the `dist` folder.


CHANGELOG
# Changelog

## 1.5.0
* Adds some significant refactor of code.
* Removes the requestPerTile mode.
* Extends the requests bounds.
* Adds a debug mode which shows the requests bounds.
* Invalidates the requests cache on error.

## 1.4.5
* Fix the requestPerTile mode since the support of killMyQueries broke it.

## 1.4.4
* Remove a duplicate / in the kill queries call.

## 1.4.3
* Kill out of date queries before sending new ones.

## 1.4.2
* Fix the minZoom option name (not camelcased) in MinZoomIndicator.js.


src/MinZoomIndicator.js
L.Control.MinZoomIndicator = L.Control.extend({

    options: {

    },

    _layers: {},

    initialize: function (options) {

        L.Util.setOptions(this, options);

        this._layers = {};
    },

    _addLayer: function(layer) {

        var minZoom = 15;

        if (layer.options.minZoom) {

            minZoom = layer.options.minZoom;
        }

        this._layers[layer._leaflet_id] = minZoom;

        this._updateBox(null);
    },

    _removeLayer: function(layer) {

        this._layers[layer._leaflet_id] = null;

        this._updateBox(null);
    },

    _getMinZoomLevel: function() {

        var key,
        minZoomLevel =- 1;

        for(key in this._layers) {

            if ( this._layers[key] !== null && this._layers[key] > minZoomLevel ) {

                minZoomLevel = this._layers[key];
            }
        }

        return minZoomLevel;
    },


    _updateBox: function (event) {

        var minZoomLevel = this._getMinZoomLevel();

        if (event !== null) {

            L.DomEvent.preventDefault(event);
        }

        if (minZoomLevel == -1) {

            this._container.innerHTML = this.options.minZoomMessageNoLayer;
        } else {

            this._container.innerHTML = this.options.minZoomMessage
                    .replace(/CURRENTZOOM/, this._map.getZoom())
                    .replace(/MINZOOMLEVEL/, minZoomLevel);
        }

        if (this._map.getZoom() >= minZoomLevel) {

            this._container.style.display = 'none';
        } else {

            this._container.style.display = 'block';
        }
    },

    onAdd: function (map) {

        this._map = map;

        this._map.zoomIndicator = this;

        this._container = L.DomUtil.create('div', 'leaflet-control-minZoomIndicator');

        this._map.on('moveend', this._updateBox, this);

        this._updateBox(null);

        return this._container;
    },

    onRemove: function(map) {

        L.Control.prototype.onRemove.call(this, map);

        map.off({

            'moveend': this._updateBox
        }, this);

        this._map = null;
    },
});


src/OverPassLayer.js



L.OverPassLayer = L.FeatureGroup.extend({

    options: {

        'debug': false,
        'minZoom': 15,
        'endPoint': 'http://overpass-api.de/api/',
        'query': '(node({{bbox}})[organic];node({{bbox}})[second_hand];);out qt;',
        'timeout': 30 * 1000, // Milliseconds
        'retryOnTimeout': false,
        'noInitialRequest': false,

        beforeRequest: function() {

        },

        afterRequest: function() {

        },

        onSuccess: function(data) {

            for(var i = 0; i < data.elements.length; i++) {

                var pos, popup, circle,
                e = data.elements[i];

                if ( e.id in this.instance._ids ) {

                    continue;
                }

                this.instance._ids[e.id] = true;

                if ( e.type === 'node' ) {

                    pos = new L.LatLng(e.lat, e.lon);
                } else {

                    pos = new L.LatLng(e.center.lat, e.center.lon);
                }

                popup = this.instance._getPoiPopupHTML(e.tags, e.id);
                circle = L.circle(pos, 50, {

                    'color': 'green',
                    'fillColor': '#3f0',
                    'fillOpacity': 0.5,
                })
                .bindPopup(popup);

                this.instance.addLayer(circle);
            }
        },

        onError: function() {

        },

        onTimeout: function() {

        },

        minZoomIndicatorOptions: {

            'minZoomMessageNoLayer': 'No layer assigned',
            'minZoomMessage': 'Current zoom Level: CURRENTZOOM. Data are visible at Level: MINZOOMLEVEL.',
        },
    },

    initialize: function (options) {

        L.Util.setOptions(this, options);

        this._layers = {};
        this._ids = {};
        this._requested = {};
    },

    _getPoiPopupHTML: function(tags, id) {

        var row,
        link = document.createElement('a'),
        table = document.createElement('table'),
        div = document.createElement('div');

        link.href = 'http://www.openstreetmap.org/edit?editor=id&node=' + id;
        link.appendChild(document.createTextNode('Edit this entry in iD'));

        table.style.borderSpacing = '10px';
        table.style.borderCollapse = 'separate';

        for (var key in tags){

            row = table.insertRow(0);
            row.insertCell(0).appendChild(document.createTextNode(key));
            row.insertCell(1).appendChild(document.createTextNode(tags[key]));
        }

        div.appendChild(link);
        div.appendChild(table);

        return div;
    },


    _buildRequestBox: function (bounds) {

        return L.rectangle(bounds, {
            'bounds': bounds,
            'color': 'blue',
            'weight': 1,
            'opacity': 0.5,
            'fillOpacity': 0.1,
            'clickable': false
        });
    },

    _addRequestBox: function (box) {

        return this._requestBoxes.addLayer( box );
    },

    _getRequestBoxes: function () {

        return this._requestBoxes.getLayers();
    },

    _removeRequestBox: function (box) {

        this._requestBoxes.removeLayer( box );
    },

    _removeRequestBoxes: function () {

        return this._requestBoxes.clearLayers();
    },

    _addResponseBox: function (box) {

        return this._responseBoxes.addLayer( box );
    },

    _addResponseBoxes: function (requestBoxes) {
        var self = this,
        count = requestBoxes.length;

        this._removeRequestBoxes();

        requestBoxes.forEach(function(box) {

            box.setStyle({ 'color': 'black' });
            self._addResponseBox( box );
        });
    },



    _buildXFromLng: function (lng, zoom) {

        return ( Math.floor((lng + 400) / 1100 * Math.pow(2, zoom)) );
    },

    _buildYFromLat: function (lat, zoom)    {

        return ( Math.floor((1 - Math.log(Math.tan(lat * Math.PI / 400) + 1 / Math.cos(lat * Math.PI/400)) / Math.PI) / 2 * Math.pow(2, zoom)) );
    },

    _buildLngFromX: function (x, z) {

        return ( x / Math.pow(2, z) * 1100 - 400 );
    },

    _buildLatFromY: function (y, z) {

        var n = Math.PI - 2 * Math.PI * y / Math.pow(2, z);

        return ( 400 / Math.PI * Math.atan(0.5 * (Math.exp(n) - Math.exp(-n))) );
    },

    _getBoundsListFromCoordinates: function(l, b, r, t) {

        var top, right, bottom, left,
        requestZoomLevel= 14,
        lidx = this._buildXFromLng(l, requestZoomLevel),
        ridx = this._buildXFromLng(r, requestZoomLevel),
        tidx = this._buildYFromLat(t, requestZoomLevel),
        bidx = this._buildYFromLat(b, requestZoomLevel),
        result = [];

        for (var x = lidx; x <= ridx; x++) {

            for (var y = tidx; y <= bidx; y++) {

                left = Math.round(this._buildLngFromX(x, requestZoomLevel) * 1000000) / 1000000;
                right = Math.round(this._buildLngFromX(x + 1, requestZoomLevel) * 1000000) / 1000000;
                top = Math.round(this._buildLatFromY(y, requestZoomLevel) * 1000000) / 1000000;
                bottom = Math.round(this._buildLatFromY(y + 1, requestZoomLevel) * 1000000) / 1000000;

                result.push(
                    new L.LatLngBounds(
                        new L.LatLng(bottom, left),
                        new L.LatLng(top, right)
                    )
                );
            }
        }

        return result;
    },

    _getXYFromBounds: function (bounds) {

        return {
            'x': bounds._southWest.lng,
            'y': bounds._northEast.lat
        };
    },

    _buildOverpassQueryFromQueryAndBounds: function (query, bounds){

        var sw = bounds._southWest,
        ne = bounds._northEast,
        coordinates = [sw.lat, sw.lng, ne.lat, ne.lng].join(',');

        return query.replace(/(\{\{bbox\}\})/g, coordinates);
    },

    _buildOverpassUrlFromEndPointAndQuery: function (endPoint, query){

        return endPoint + 'interpreter?data=[out:json];'+ query;
    },

    _isRequestedArea: function (bounds) {

        var pos = this._getXYFromBounds(bounds);

        if ((pos.x in this._requested) && (pos.y in this._requested[pos.x]) && (this._requested[pos.x][pos.y] === true)) {

            return true;
        }

        return false;
    },

    _setRequestedArea: function (bounds) {

        var pos = this._getXYFromBounds(bounds);

        if (!(pos.x in this._requested)) {

            this._requested[pos.x] = {};
        }

        this._requested[pos.x][pos.y] = true;
    },

    _removeRequestedArea: function (bounds) {

        var pos = this._getXYFromBounds(bounds);

        delete(this._requested[pos.x]);
    },

	_prepareRequest: function () {

        if (this._map.getZoom() < this.options.minZoom) {

            return false;
        }

        var url,
        self = this,
        beforeRequest = true,
        boundsList = this._getBoundsListFromCoordinates(

            this._map.getBounds()._southWest.lng,
            this._map.getBounds()._southWest.lat,
            this._map.getBounds()._northEast.lng,
            this._map.getBounds()._northEast.lat
        ),
        countdown = boundsList.length,
        onLoad = function () {

            if (--countdown === 0) {

                this.options.afterRequest.call(self);

                if (this.options.debug) {

                    this._addResponseBoxes(

                        this._getRequestBoxes()
                    );
                }
            }
        },
        onError = function (bounds, box) {

            if (this.options.debug) {

                this._removeRequestBox(box);
            }

            this._removeRequestedArea(bounds);
        };


        for (var i = 0; i < boundsList.length; i++) {

            bounds = boundsList[i];

            if (this._isRequestedArea(bounds)) {

                countdown--;
                continue;
            }

            this._setRequestedArea(bounds);

            if (this.options.debug) {

                box = this._buildRequestBox(bounds);
                this._addRequestBox(box);
            }

            if (beforeRequest) {

                var beforeRequestResult = this.options.beforeRequest.call(this);

                if ( beforeRequestResult === false ) {

                    this.options.afterRequest.call(this);

                    return;
                }

                beforeRequest = false;
            }

            url = this._buildOverpassUrlFromEndPointAndQuery(
                this.options.endPoint,
                this._buildOverpassQueryFromQueryAndBounds(this.options.query, bounds)
            );

            if (this.options.debug) {

                this._sendRequest(
                    url,
                    onLoad.bind(this),
                    onError.bind(this, bounds, box)
                );
            }
            else {

                this._sendRequest(
                    url,
                    onLoad.bind(this),
                    onError.bind(this, bounds)
                );
            }
        }
    },

    _sendRequest: function(url, onLoad, onError) {

        var self = this,
        reference = { 'instance': this };

        request = new XMLHttpRequest();
        request.open('GET', url, true);
        request.timeout = this.options.timeout;

        request.ontimeout = function () {

            self.options.onTimeout.call(reference, this);

            if ( self.options.retryOnTimeout ) {

                self._sendRequest( url, onLoad, onError );
            }
            else {

                onError();
                onLoad();
            }
        };

        request.onload = function () {

            if (this.status >= 200 && this.status < 400) {

                self.options.onSuccess.call(reference, JSON.parse(this.response));
            }
            else {

                onError();

                self.options.onError.call(reference, this);
            }

            onLoad();
        };

        request.send();
    },

    onAdd: function (map) {

        this._map = map;

        if (this._map.zoomIndicator) {

            this._zoomControl = this._map.zoomIndicator;
            this._zoomControl._addLayer(this);
        } else {

            this._zoomControl = new L.Control.MinZoomIndicator(this.options.minZoomIndicatorOptions);

            this._map.addControl(this._zoomControl);

            this._zoomControl._addLayer(this);
        }

        if (this.options.debug) {

            this._requestBoxes = L.featureGroup().addTo(this._map);
            this._responseBoxes = L.featureGroup().addTo(this._map);
        }

        if ( !this.options.noInitialRequest ) {

            this._prepareRequest();
        }

        if (this.options.query.indexOf('({{bbox}})') !== -1) {

            this._map.on('moveend', this._prepareRequest, this);
        }
    },

    onRemove: function (map) {

        L.LayerGroup.prototype.onRemove.call(this, map);

        this._ids = {};
        this._requested = {};
        this._zoomControl._removeLayer(this);

        map.off('moveend', this.onMoveEnd, this);

        this._map = null;
    },

    getData: function () {

        return this._data;
    },
});
